File: packages/format/test/av1-sequence-header.test.ts
import { describe, expect, it } from "vitest";

import { parseAv1SequenceHeader } from "../src/av1/sequence-header.js";

const LIBAOM_64_X_32_8_BIT = Uint8Array.from(
  Buffer.from("00000002a7ff36be4404040410", "hex")
);

describe("AV1 sequence-header parsing", () => {
  it("parses the libaom Main-profile BT.709 sequence header", () => {
    expect(parseAv1SequenceHeader(LIBAOM_64_X_32_8_BIT)).toMatchObject({
      profile: 0,
      level: 0,
      tier: "M",
      bitDepth: 8,
      maxWidth: 64,
      maxHeight: 32,
      monochrome: false,
      subsamplingX: 1,
      subsamplingY: 1,
      colorPrimaries: 1,
      transferCharacteristics: 1,
      matrixCoefficients: 1,
      fullRange: false
    });
  });

  it("rejects truncation and color-description mutation", () => {
    expect(() => parseAv1SequenceHeader(LIBAOM_64_X_32_8_BIT.slice(0, 4)))
      .toThrow(/truncated/u);
    const changed = LIBAOM_64_X_32_8_BIT.slice();
    changed[10] = (changed[10] ?? 0) ^ 0x80;
    expect(() => parseAv1SequenceHeader(changed)).toThrow();
  });
});

File: flutter/packages/aval_format/test/av1_sequence_header_test.dart
/// Dart port of `packages/format/test/av1-sequence-header.test.ts`.
library;

import 'dart:typed_data';

import 'package:aval_format/src/av1/index.dart';
import 'package:aval_format/src/errors.dart';
import 'package:test/test.dart';

// Hex "00000002a7ff36be4404040410".
final Uint8List libaom64x32_8bit = Uint8List.fromList([
  0x00, 0x00, 0x00, 0x02, 0xa7, 0xff, 0x36, 0xbe, 0x44, 0x04, 0x04, 0x04, 0x10,
]);

void main() {
  group('AV1 sequence-header parsing', () {
    test('parses the libaom Main-profile BT.709 sequence header', () {
      final sequence = parseAv1SequenceHeader(libaom64x32_8bit);
      expect(sequence.profile, 0);
      expect(sequence.level, 0);
      expect(sequence.tier, 'M');
      expect(sequence.bitDepth, 8);
      expect(sequence.maxWidth, 64);
      expect(sequence.maxHeight, 32);
      expect(sequence.monochrome, isFalse);
      expect(sequence.subsamplingX, 1);
      expect(sequence.subsamplingY, 1);
      expect(sequence.colorPrimaries, 1);
      expect(sequence.transferCharacteristics, 1);
      expect(sequence.matrixCoefficients, 1);
      expect(sequence.fullRange, isFalse);
    });

    test('rejects truncation and color-description mutation', () {
      expect(
        () => parseAv1SequenceHeader(libaom64x32_8bit.sublist(0, 4)),
        throwsA(predicate<Object>((error) =>
            error is FormatError &&
            RegExp('truncated').hasMatch(error.message))),
      );
      final changed = Uint8List.fromList(libaom64x32_8bit);
      changed[10] = changed[10] ^ 0x80;
      expect(() => parseAv1SequenceHeader(changed), throwsA(isA<FormatError>()));
    });
  });
}
